File: ios/SpeakLab/Services/Coach.swift
import Foundation

/// Orchestrates coaching: build the prompt, call the service, validate the
/// response, check it against the transcript, and fall back to the on-device
/// engine when any of that fails.
///
/// Callers never see a failure. They get feedback plus honest provenance —
/// whether it came from Claude or from the device, and whether the quoted
/// evidence was actually found in what the learner said.
@MainActor
final class Coach: ObservableObject {

    struct FeedbackResult {
        var feedback: CoachFeedback
        /// True when this came from the on-device engine rather than the model.
        var wasLocal: Bool
        /// False when the evidence quote could not be located in the transcript.
        var evidenceVerified: Bool
        /// Shown to the learner when something degraded, so nothing is silent.
        var note: String?
    }

    struct ComparisonResult {
        var comparison: AttemptComparison
        var wasLocal: Bool
        var note: String?
    }

    private let serviceProvider: @MainActor () -> CoachingService

    init(serviceProvider: @escaping @MainActor () -> CoachingService) {
        self.serviceProvider = serviceProvider
    }

    var isConnected: Bool { serviceProvider().isConfigured }

    // MARK: - Speaking attempt

    func analyzeAttempt(
        scenario: Scenario,
        skill: MicroSkill,
        transcript: String,
        metrics: SpeakingMetrics,
        attemptNumber: Int,
        previousTargetSkillID: String? = nil
    ) async -> FeedbackResult {
        let localFallback = {
            LocalFeedbackEngine.feedback(
                scenario: scenario,
                skill: skill,
                transcript: transcript,
                metrics: metrics
            )
        }

        guard !transcript.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return FeedbackResult(
                feedback: localFallback(),
                wasLocal: true,
                evidenceVerified: true,
                note: "No speech was recognised, so this is based on the recording's timing only."
            )
        }

        let request = PromptBuilder.attemptFeedbackRequest(
            scenario: scenario,
            skill: skill,
            transcript: transcript,
            metrics: metrics,
            attemptNumber: attemptNumber,
            previousTargetSkillID: previousTargetSkillID
        )

        do {
            let data = try await serviceProvider().run(request)
            let feedback = try CoachingDecoder.decodeFeedback(from: data)
            let grounding = TranscriptGrounding.verify(quote: feedback.evidenceQuote, in: transcript)

            // The quote is kept either way; `evidenceVerified` is what stops the
            // UI presenting an unverified one as a verbatim record.
            return FeedbackResult(
                feedback: feedback,
                wasLocal: false,
                evidenceVerified: grounding.verified,
                note: grounding.verified ? nil : "The quoted moment is a paraphrase — it wasn't found word-for-word in your transcript."
            )
        } catch {
            return FeedbackResult(
                feedback: localFallback(),
                wasLocal: true,
                evidenceVerified: true,
                note: Self.degradedNote(for: error)
            )
        }
    }

    // MARK: - Conversation debrief

    func debriefConversation(
        scenario: Scenario,
        skill: MicroSkill,
        turns: [ConversationTurn],
        metrics: ConversationMetrics,
        attemptNumber: Int
    ) async -> FeedbackResult {
        let localFallback = {
            LocalFeedbackEngine.conversationFeedback(
                scenario: scenario,
                skill: skill,
                metrics: metrics,
                turns: turns
            )
        }

        let request = PromptBuilder.conversationDebriefRequest(
            scenario: scenario,
            skill: skill,
            turns: turns,
            metrics: metrics,
            attemptNumber: attemptNumber
        )

        let learnerText = turns.filter { $0.speaker == .user }.map { $0.text }.joined(separator: " ")

        do {
            let data = try await serviceProvider().run(request)
            let feedback = try CoachingDecoder.decodeFeedback(from: data)
            let grounding = TranscriptGrounding.verify(quote: feedback.evidenceQuote, in: learnerText)
            return FeedbackResult(
                feedback: feedback,
                wasLocal: false,
                evidenceVerified: grounding.verified,
                note: grounding.verified ? nil : "The quoted moment is a paraphrase rather than your exact words."
            )
        } catch {
            return FeedbackResult(
                feedback: localFallback(),
                wasLocal: true,
                evidenceVerified: true,
                note: Self.degradedNote(for: error)
            )
        }
    }

    // MARK: - Retry comparison

    func compareAttempts(
        scenario: Scenario,
        skill: MicroSkill,
        target: String,
        targetSkillID: String?,
        retryInstruction: String,
        first: (transcript: String, metrics: SpeakingMetrics),
        second: (transcript: String, metrics: SpeakingMetrics)
    ) async -> ComparisonResult {
        let request = PromptBuilder.comparisonRequest(
            scenario: scenario,
            skill: skill,
            target: target,
            retryInstruction: retryInstruction,
            firstTranscript: first.transcript,
            firstMetrics: first.metrics,
            secondTranscript: second.transcript,
            secondMetrics: second.metrics
        )

        do {
            let data = try await serviceProvider().run(request)
            let comparison = try CoachingDecoder.decodeComparison(from: data)
            return ComparisonResult(comparison: comparison, wasLocal: false, note: nil)
        } catch {
            return ComparisonResult(
                comparison: LocalFeedbackEngine.compare(
                    target: target,
                    targetSkillID: targetSkillID,
                    first: first.metrics,
                    second: second.metrics
                ),
                wasLocal: true,
                note: Self.degradedNote(for: error)
            )
        }
    }

    // MARK: - Conversation turns

    /// The simulated person's next line.
    ///
    /// Unlike feedback, this has no honest offline equivalent from the model, so
    /// the caller is told when a scripted stand-in was used.
    func characterTurn(
        scenario: Scenario,
        history: [ConversationTurn],
        turnBudget: Int
    ) async -> (turn: CharacterTurnResponse, wasScripted: Bool) {
        let request = PromptBuilder.characterTurnRequest(
            scenario: scenario,
            history: history,
            turnBudget: turnBudget
        )
        do {
            let data = try await serviceProvider().run(request)
            let turn = try CoachingDecoder.decodeCharacterTurn(from: data)
            return (turn, false)
        } catch {
            return (ScriptedCharacter.reply(scenario: scenario, history: history, turnBudget: turnBudget), true)
        }
    }

    // MARK: - Helpers

    private static func degradedNote(for error: Error) -> String {
        if let serviceError = error as? CoachingServiceError {
            switch serviceError {
            case .notConfigured:
                return "Coached on device. Connect a coaching server in Settings for deeper feedback."
            case .offline:
                return "You're offline, so this was coached on device."
            case .refused:
                return "The coach declined to analyse this one, so this is on-device feedback."
            default:
                return "\(serviceError.localizedDescription) Coached on device instead."
            }
        }
        if error is CoachingDecodingError {
            return "The coach's reply didn't arrive in one piece, so this was coached on device."
        }
        return "Coached on device."
    }
}

/// A small scripted stand-in used when no coaching server is configured.
///
/// It is not trying to be Claude. It exists so the whole training loop —
/// including conversation mode — can be walked through on a device with no
/// setup at all, and it says plainly in the UI that it is a rehearsal.
enum ScriptedCharacter {

    static func reply(
        scenario: Scenario,
        history: [ConversationTurn],
        turnBudget: Int
    ) -> CharacterTurnResponse {
        guard let character = scenario.character else {
            return CharacterTurnResponse(speech: "Okay.", innerState: "", shouldEnd: true)
        }

        let userTurns = history.filter { $0.speaker == .user }

        if history.isEmpty {
            return CharacterTurnResponse(speech: character.opensWith, innerState: character.emotionalState)
        }

        let lastUserText = userTurns.last?.text ?? ""
        let questions = ConversationMetricsCalculator.detectQuestions(in: lastUserText, turnIndex: 0)
        let askedOpen = questions.contains { $0.kind == .open }

        // Enough exchanges have happened — wrap up.
        if userTurns.count >= max(3, turnBudget - 2) {
            return CharacterTurnResponse(
                speech: "Alright — let me have a think about it and I'll let you know.",
                innerState: "Neither convinced nor put off.",
                objectiveMet: false,
                objectiveMissed: false,
                shouldEnd: true,
                endReason: "The conversation reached its natural end."
            )
        }

        // A real question earns more information; anything else gets the objection again.
        if askedOpen {
            return CharacterTurnResponse(
                speech: opennessLine(for: character, turn: userTurns.count),
                innerState: "Starting to feel listened to.",
                observedMove: "Asked an open question."
            )
        }

        return CharacterTurnResponse(
            speech: guardedLine(for: character, turn: userTurns.count),
            innerState: "Not sure this person is listening yet.",
            observedMove: questions.isEmpty ? "Explained rather than asked." : "Asked a closed question."
        )
    }

    private static func opennessLine(for character: CharacterBrief, turn: Int) -> String {
        let lines = [
            "Honestly? \(character.hiddenGoal.prefix(1).lowercased() + character.hiddenGoal.dropFirst())",
            "I suppose what's really behind it is this — \(character.objection)",
            "That's a fair question. I hadn't really put it into words before."
        ]
        return lines[min(turn, lines.count - 1)]
    }

    private static func guardedLine(for character: CharacterBrief, turn: Int) -> String {
        let lines = [
            character.objection,
            "Right. I hear you, but it's still the same issue for me.",
            "Mm. I'm not sure that changes much, to be honest."
        ]
        return lines[min(turn, lines.count - 1)]
    }
}

File: web/src/services/coach.ts
import {
  attemptFeedbackRequest,
  characterTurnRequest,
  comparisonRequest,
  conversationDebriefRequest,
  type CoachingRequest,
} from '../core/prompts'
import {
  CoachingDecodingError,
  decodeCharacterTurn,
  decodeComparison,
  decodeFeedback,
} from '../core/decoder'
import { verifyQuote } from '../core/grounding'
import { detectQuestions } from '../core/conversationMetrics'
import { localCompare, localConversationFeedback, localSpeakingFeedback } from '../core/localCoach'
import type {
  AttemptComparison,
  CharacterTurnResponse,
  CoachFeedback,
  ConversationMetrics,
  ConversationTurn,
  MicroSkill,
  Scenario,
  SpeakingMetrics,
} from '../core/types'

/**
 * The coaching facade.
 *
 * Everything the app asks of a model goes through here, and every path has a
 * local fallback. That is deliberate: a page served from static hosting has no
 * server of its own, so "no coaching service configured" is the *normal* state,
 * not an error state. The app has to be fully usable in it.
 *
 * The API key never reaches this code. When a proxy is configured it holds the
 * key; the browser only ever sees a URL and an optional shared secret that the
 * learner typed in themselves.
 */

export interface CoachSettings {
  /** Base URL of a deployed `server/` instance, e.g. https://coach.example.com */
  proxyURL?: string
  /** Optional shared secret the proxy requires in `x-speaklab-key`. */
  sharedSecret?: string
}

export interface FeedbackOutcome {
  feedback: CoachFeedback
  wasLocal: boolean
  evidenceVerified: boolean
  note?: string
}

export interface ComparisonOutcome {
  comparison: AttemptComparison
  wasLocal: boolean
  note?: string
}

export interface CharacterOutcome {
  turn: CharacterTurnResponse
  wasScripted: boolean
  note?: string
}

export class CoachingServiceError extends Error {
  readonly code: string
  constructor(code: string, message: string) {
    super(message)
    this.name = 'CoachingServiceError'
    this.code = code
  }
}

export function isConfigured(settings: CoachSettings): boolean {
  return Boolean(settings.proxyURL && settings.proxyURL.trim().length > 0)
}

/** Posts one coaching request and returns the model's raw JSON text. */
async function callProxy(request: CoachingRequest, settings: CoachSettings): Promise<string> {
  const base = settings.proxyURL?.trim().replace(/\/+$/, '')
  if (!base) throw new CoachingServiceError('notConfigured', 'No coaching server configured.')

  const headers: Record<string, string> = { 'content-type': 'application/json' }
  if (settings.sharedSecret) headers['x-speaklab-key'] = settings.sharedSecret

  let response: Response
  try {
    response = await fetch(`${base}/v1/coach`, {
      method: 'POST',
      headers,
      body: JSON.stringify(request),
    })
  } catch {
    throw new CoachingServiceError(
      'offline',
      'The coaching server could not be reached from this browser.',
    )
  }

  if (!response.ok) {
    const detail = (await response.json().catch(() => undefined)) as
      | { error?: { code?: string; message?: string } }
      | undefined
    const code = detail?.error?.code ?? String(response.status)
    if (response.status === 401) {
      throw new CoachingServiceError('unauthorized', 'The coaching server rejected the shared key.')
    }
    if (response.status === 422) {
      throw new CoachingServiceError('refused', 'The coach declined to analyse this one.')
    }
    if (response.status === 429) {
      throw new CoachingServiceError('rateLimited', 'The coaching server is rate limiting requests.')
    }
    throw new CoachingServiceError(code, detail?.error?.message ?? 'The coaching server errored.')
  }

  const payload = (await response.json()) as { data?: unknown }
  if (typeof payload.data === 'string') return payload.data
  if (payload.data && typeof payload.data === 'object') return JSON.stringify(payload.data)
  throw new CoachingServiceError('malformed', 'The coaching server returned no data.')
}

function degradedNote(error: unknown): string {
  if (error instanceof CoachingServiceError) {
    switch (error.code) {
      case 'notConfigured':
        return 'Coached in your browser. Connect a coaching server in Settings for deeper feedback.'
      case 'offline':
        return "The coaching server couldn't be reached, so this was coached in your browser."
      case 'refused':
        return 'The coach declined to analyse this one, so this is browser-only feedback.'
      case 'unauthorized':
        return 'The coaching server rejected the shared key, so this was coached in your browser.'
      default:
        return `${error.message} Coached in your browser instead.`
    }
  }
  if (error instanceof CoachingDecodingError) {
    return "The coach's reply didn't arrive in one piece, so this was coached in your browser."
  }
  return 'Coached in your browser.'
}

// MARK: - Speaking

export async function analyzeAttempt(options: {
  scenario: Scenario
  skill: MicroSkill
  transcript: string
  metrics: SpeakingMetrics
  attemptNumber: number
  settings: CoachSettings
}): Promise<FeedbackOutcome> {
  const { scenario, skill, transcript, metrics, attemptNumber, settings } = options
  const local = () => localSpeakingFeedback(scenario, skill, metrics)

  if (!isConfigured(settings)) {
    return {
      feedback: local(),
      wasLocal: true,
      evidenceVerified: true,
      note: degradedNote(new CoachingServiceError('notConfigured', '')),
    }
  }

  try {
    const raw = await callProxy(
      attemptFeedbackRequest({ scenario, skill, transcript, metrics, attemptNumber }),
      settings,
    )
    const feedback = decodeFeedback(raw)
    const grounding = verifyQuote(feedback.evidenceQuote, transcript)
    const outcome: FeedbackOutcome = {
      feedback,
      wasLocal: false,
      evidenceVerified: grounding.verified,
    }
    if (!grounding.verified) {
      outcome.note = "The quoted evidence isn't a word-for-word match, so it's marked as a paraphrase."
    }
    return outcome
  } catch (error) {
    return { feedback: local(), wasLocal: true, evidenceVerified: true, note: degradedNote(error) }
  }
}

export async function compareAttempts(options: {
  scenario: Scenario
  skill: MicroSkill
  target: string
  targetSkillID: string
  retryInstruction: string
  first: { transcript: string; metrics: SpeakingMetrics }
  second: { transcript: string; metrics: SpeakingMetrics }
  settings: CoachSettings
}): Promise<ComparisonOutcome> {
  const { scenario, skill, target, targetSkillID, retryInstruction, first, second, settings } =
    options
  const local = () => localCompare(targetSkillID, first.metrics, second.metrics)

  if (!isConfigured(settings)) {
    return {
      comparison: local(),
      wasLocal: true,
      note: degradedNote(new CoachingServiceError('notConfigured', '')),
    }
  }

  try {
    const raw = await callProxy(
      comparisonRequest({
        scenario,
        skill,
        target,
        retryInstruction,
        firstTranscript: first.transcript,
        firstMetrics: first.metrics,
        secondTranscript: second.transcript,
        secondMetrics: second.metrics,
      }),
      settings,
    )
    return { comparison: decodeComparison(raw), wasLocal: false }
  } catch (error) {
    return { comparison: local(), wasLocal: true, note: degradedNote(error) }
  }
}

// MARK: - Conversation

export async function characterTurn(options: {
  scenario: Scenario
  history: ConversationTurn[]
  turnBudget: number
  settings: CoachSettings
}): Promise<CharacterOutcome> {
  const { scenario, history, turnBudget, settings } = options
  const local = () => scriptedReply(scenario, history, turnBudget)

  if (!isConfigured(settings)) {
    return { turn: local(), wasScripted: true }
  }

  try {
    const raw = await callProxy(characterTurnRequest({ scenario, history, turnBudget }), settings)
    return { turn: decodeCharacterTurn(raw), wasScripted: false }
  } catch (error) {
    return { turn: local(), wasScripted: true, note: degradedNote(error) }
  }
}

export async function debriefConversation(options: {
  scenario: Scenario
  skill: MicroSkill
  turns: ConversationTurn[]
  metrics: ConversationMetrics
  attemptNumber: number
  settings: CoachSettings
}): Promise<FeedbackOutcome> {
  const { scenario, skill, turns, metrics, attemptNumber, settings } = options
  const local = () => localConversationFeedback(skill, metrics, turns)
  const learnerText = turns
    .filter((turn) => turn.speaker === 'user')
    .map((turn) => turn.text)
    .join(' ')

  if (!isConfigured(settings)) {
    return {
      feedback: local(),
      wasLocal: true,
      evidenceVerified: true,
      note: degradedNote(new CoachingServiceError('notConfigured', '')),
    }
  }

  try {
    const raw = await callProxy(
      conversationDebriefRequest({ scenario, skill, turns, metrics, attemptNumber }),
      settings,
    )
    const feedback = decodeFeedback(raw)
    const grounding = verifyQuote(feedback.evidenceQuote, learnerText)
    const outcome: FeedbackOutcome = {
      feedback,
      wasLocal: false,
      evidenceVerified: grounding.verified,
    }
    if (!grounding.verified) {
      outcome.note = "The quoted evidence isn't a word-for-word match, so it's marked as a paraphrase."
    }
    return outcome
  } catch (error) {
    return { feedback: local(), wasLocal: true, evidenceVerified: true, note: degradedNote(error) }
  }
}

/**
 * A small scripted stand-in used when no coaching server is configured.
 *
 * It is not trying to be Claude. It exists so the whole training loop —
 * including conversation mode — can be walked through with no setup at all, and
 * the UI says plainly that it is a rehearsal rather than a real simulation.
 */
export function scriptedReply(
  scenario: Scenario,
  history: ConversationTurn[],
  turnBudget: number,
): CharacterTurnResponse {
  const character = scenario.character
  if (!character) {
    return { speech: 'Okay.', innerState: '', objectiveMet: false, objectiveMissed: false, shouldEnd: true }
  }

  if (history.length === 0) {
    return {
      speech: character.opensWith,
      innerState: character.emotionalState,
      objectiveMet: false,
      objectiveMissed: false,
      shouldEnd: false,
    }
  }

  const userTurns = history.filter((turn) => turn.speaker === 'user')
  const lastUserText = userTurns[userTurns.length - 1]?.text ?? ''
  const questions = detectQuestions(lastUserText, 0)
  const askedOpen = questions.some((question) => question.kind === 'open')

  if (userTurns.length >= Math.max(3, turnBudget - 2)) {
    return {
      speech: "Alright — let me have a think about it and I'll let you know.",
      innerState: 'Neither convinced nor put off.',
      objectiveMet: false,
      objectiveMissed: false,
      shouldEnd: true,
      endReason: 'The conversation reached its natural end.',
    }
  }

  if (askedOpen) {
    return {
      speech: opennessLine(character.hiddenGoal, character.objection, userTurns.length),
      innerState: 'Starting to feel listened to.',
      observedMove: 'Asked an open question.',
      objectiveMet: false,
      objectiveMissed: false,
      shouldEnd: false,
    }
  }

  return {
    speech: guardedLine(character.objection, userTurns.length),
    innerState: 'Not sure this person is listening yet.',
    observedMove: questions.length === 0 ? 'Explained rather than asked.' : 'Asked a closed question.',
    objectiveMet: false,
    objectiveMissed: false,
    shouldEnd: false,
  }
}

function opennessLine(hiddenGoal: string, objection: string, turn: number): string {
  const lines = [
    `Honestly? ${hiddenGoal.charAt(0).toLowerCase()}${hiddenGoal.slice(1)}`,
    `I suppose what's really behind it is this — ${objection}`,
    "That's a fair question. I hadn't really put it into words before.",
  ]
  return lines[Math.min(turn, lines.length - 1)] as string
}

function guardedLine(objection: string, turn: number): string {
  const lines = [
    objection,
    "Right. I hear you, but it's still the same issue for me.",
    "Mm. I'm not sure that changes much, to be honest.",
  ]
  return lines[Math.min(turn, lines.length - 1)] as string
}
